File: src/cmdbox/services/history_service.py
import json
from typing import Callable

from cmdbox.models import CommandHistory, Profile
from cmdbox.repositories.history_repository import HistoryRepository
from cmdbox.repositories.profile_repository import ProfileRepository
from cmdbox.services.errors import HistoryIndexError
from cmdbox.settings.models import Settings


class HistoryService:

    def __init__(
        self,
        repo: HistoryRepository,
        get_settings: Callable[[], Settings],
        profile_repository: ProfileRepository,
    ):
        self._repo = repo
        self._get_settings = get_settings
        self._profile_repo = profile_repository

    def resolve_profile(self, profile: str | None) -> Profile | None:
        return self._profile_repo.get_by_name(profile) if profile else None

    def get_recent(
        self,
        alias: str | None = None,
        limit: int = 20,
        profile: str | None = None,
    ) -> list[CommandHistory]:
        resolved_profile = self.resolve_profile(profile)
        return self._repo.get_recent(alias, limit, profile=resolved_profile)

    def get_by_ref(
        self,
        ref: str,
        alias: str | None = None,
        profile: str | None = None,
    ) -> CommandHistory:
        resolved_profile = self.resolve_profile(profile)
        if ref.isdigit():
            return self._get_by_index(int(ref), alias=alias, profile=resolved_profile)
        return self._repo.get_by_id(ref, profile=resolved_profile)

    def get_variables(self, entry: CommandHistory) -> dict[str, str] | None:
        if entry.variables_used is None:
            return None
        return json.loads(entry.variables_used)

    def delete_by_ref(self, ref: str, profile: str | None = None) -> bool:
        resolved_profile = self.resolve_profile(profile)
        entry = self.get_by_ref(ref, profile=profile)
        return self._repo.delete_by_id(entry.id, profile=resolved_profile)

    def clear(
        self,
        alias: str | None = None,
        profile: str | None = None,
    ) -> int:
        resolved_profile = self.resolve_profile(profile)
        return self._repo.clear(alias=alias, profile=resolved_profile)

    def _get_by_index(
        self,
        index: int,
        alias: str | None = None,
        profile: Profile | None = None,
    ) -> CommandHistory:
        """
        Fetches a specific command history entry by its index.

        This method retrieves a command history entry from the repository based on
        the passed index. If the index is out of range or invalid, a
        HistoryIndexError is raised. The method optionally allows limiting the
        retrieval to a specific alias.

        Args:
            index (int): The one-based index of the command history entry to
                retrieve. Must be greater than 0.
            alias (str | None): Optional alias to filter the history entries. If
                None, no alias filter is applied.
            profile (str | None): Optional profile to filter the history entries. If
                None, no profile filter is applied.

        Returns:
            CommandHistory: The command history entry corresponding to the provided
            index.

        Raises:
            HistoryIndexError: If the provided index is less than 1 or if the index
            exceeds the number of available entries.
        """
        if index < 1:
            raise HistoryIndexError(index=index)
        entries = self._repo.get_recent(alias=alias, limit=index, profile=profile)
        if index > len(entries):
            raise HistoryIndexError(index=index)
        return entries[index - 1]

File: internal/services/history_service.go
package services

import (
	"encoding/json"
	"fmt"
	"strconv"

	"github.com/PhantomLambSoft/CmdBox/internal/models"
	"github.com/PhantomLambSoft/CmdBox/internal/repository"
)

type HistoryService interface {
	GetRecent(alias *string, limit *int, profileName *string) ([]models.CommandHistory, error)
	GetByRef(ref string, alias *string, profileName *string) (*models.CommandHistory, error)
	GetVariables(entry *models.CommandHistory) (map[string]string, error)
	DeleteByRef(ref string, alias *string, profileName *string) error
	Clear(alias *string, profileName *string) (int, error)
}

type historyService struct {
	historyRepo repository.HistoryRepository
	profileRepo repository.ProfileRepository
}

func NewHistoryService(historyRepo repository.HistoryRepository, profileRepo repository.ProfileRepository) HistoryService {
	return &historyService{
		historyRepo: historyRepo,
		profileRepo: profileRepo,
	}
}

// getProfileID resolves the profile ID by name or retrieves the active profile's ID if no name is provided.
func (s *historyService) getProfileID(profileName *string) (*uint, error) {
	var profile *models.Profile
	var err error
	if profileName != nil {
		profile, err = s.profileRepo.GetByName(*profileName)
		if err != nil {
			return nil, fmt.Errorf("getting active profile: %w", err)
		}
	} else {
		profile, err = s.profileRepo.GetActiveCommandProfile()
		if err != nil {
			return nil, fmt.Errorf("getting active profile: %w", err)
		}
	}
	return &profile.ID, nil
}

func (s *historyService) GetRecent(alias *string, limit *int, profileName *string) ([]models.CommandHistory, error) {
	profileID, err := s.getProfileID(profileName)
	if err != nil {
		return nil, fmt.Errorf("getting profile ID: %w", err)
	}
	resolvedLimit := 0
	if limit != nil {
		resolvedLimit = *limit
	}
	return s.historyRepo.GetRecent(alias, resolvedLimit, profileID)
}

func (s *historyService) GetByRef(ref string, alias *string, profileName *string) (*models.CommandHistory, error) {
	if index, err := strconv.Atoi(ref); err == nil {
		return s.getByIndex(index, alias, profileName)
	}

	profileID, err := s.getProfileID(profileName)
	if err != nil {
		return nil, fmt.Errorf("getting profile ID: %w", err)
	}

	return s.historyRepo.GetByID(ref, profileID)
}

func (s *historyService) GetVariables(entry *models.CommandHistory) (map[string]string, error) {
	if entry.VariablesUsed == nil {
		return nil, nil
	}

	var variables map[string]string
	if err := json.Unmarshal([]byte(*entry.VariablesUsed), &variables); err != nil {
		return nil, fmt.Errorf("unmarshaling variables: %w", err)
	}
	return variables, nil
}

func (s *historyService) DeleteByRef(ref string, alias *string, profileName *string) error {
	entry, err := s.GetByRef(ref, alias, profileName)
	if err != nil {
		return fmt.Errorf("getting history entry: %w", err)
	}

	profileID, err := s.getProfileID(profileName)
	if err != nil {
		return fmt.Errorf("getting profile ID: %w", err)
	}

	return s.historyRepo.DeleteByID(entry.ID, profileID)
}

func (s *historyService) Clear(alias *string, profileName *string) (int, error) {
	profileID, err := s.getProfileID(profileName)
	if err != nil {
		return 0, fmt.Errorf("getting profile ID: %w", err)
	}

	return s.historyRepo.Clear(alias, profileID)
}

// GetByIndex retrieves the nth most recent command history entry, based on the given index, alias, and profile name.
// Returns ErrUnknownHistoryIndex if the index is invalid or out of bounds.
func (s *historyService) getByIndex(index int, alias *string, profileName *string) (*models.CommandHistory, error) {
	if index < 1 {
		return nil, ErrUnknownHistoryIndex
	}
	entries, err := s.GetRecent(alias, &index, profileName)
	if err != nil {
		return nil, fmt.Errorf("getting recent history: %w", err)
	}
	if index > len(entries) {
		return nil, ErrUnknownHistoryIndex
	}
	return &entries[index-1], nil
}
